code/game/machinery/vendors/vending.dm

```diff
@@ -0,0 +1,880 @@
+	for (var/datum/data/vending_product/R in product_records + coin_records + hidden_records)
+		data["stock"][R.name] = R.amount
+	data["extended_inventory"] = extended_inventory
+	data["vend_ready"] = vend_ready
+	data["coin_name"] = coin ? coin.name : FALSE
+	data["panel_open"] = panel_open ? TRUE : FALSE
+	data["speaker"] = shut_up ? FALSE : TRUE
+	data["item_slot"] = item_slot // boolean
+	data["inserted_item_name"] = inserted_item ? inserted_item.name : FALSE
+	return data
+
+
+/obj/machinery/economy/vending/ui_static_data(mob/user)
+	var/list/data = list()
+	data["chargesMoney"] = length(prices) > 0 ? TRUE : FALSE
+	data["product_records"] = list()
+	var/i = 1
+	for(var/datum/data/vending_product/R in product_records)
+		var/list/data_pr = list(
+			path = replacetext(replacetext("[R.product_path]", "/obj/item/", ""), "/", "-"),
+			name = R.name,
+			price = (R.product_path in prices) ? prices[R.product_path] : 0,
+			max_amount = R.max_amount,
+			req_coin = FALSE,
+			is_hidden = FALSE,
+			inum = i
+		)
+		data["product_records"] += list(data_pr)
+		i++
+	data["coin_records"] = list()
+	for (var/datum/data/vending_product/R in coin_records)
+		var/list/data_cr = list(
+			path = replacetext(replacetext("[R.product_path]", "/obj/item/", ""), "/", "-"),
+			name = R.name,
+			price = (R.product_path in prices) ? prices[R.product_path] : 0,
+			max_amount = R.max_amount,
+			req_coin = TRUE,
+			is_hidden = FALSE,
+			inum = i,
+			premium = TRUE
+		)
+		data["coin_records"] += list(data_cr)
+		i++
+	data["hidden_records"] = list()
+	for (var/datum/data/vending_product/R in hidden_records)
+		var/list/data_hr = list(
+			path = replacetext(replacetext("[R.product_path]", "/obj/item/", ""), "/", "-"),
+			name = R.name,
+			price = (R.product_path in prices) ? prices[R.product_path] : 0,
+			max_amount = R.max_amount,
+			req_coin = FALSE,
+			is_hidden = TRUE,
+			inum = i,
+			premium = TRUE
+		)
+		data["hidden_records"] += list(data_hr)
+		i++
+	data["imagelist"] = imagelist
+	return data
+
+/obj/machinery/economy/vending/ui_act(action, params, datum/tgui/ui)
+	. = ..()
+	if(.)
+		return
+
+	var/mob/living/user = ui.user
+
+	if(issilicon(user) && !isrobot(user))
+		to_chat(user, "<span class='warning'>The vending machine refuses to interface with you, as you are not in its target demographic!</span>")
+		return
+	switch(action)
+		if("toggle_voice")
+			if(panel_open)
+				shut_up = !shut_up
+				. = TRUE
+		if("eject_item")
+			eject_item(user)
+			. = TRUE
+		if("remove_coin")
+			if(!coin)
+				to_chat(user, "<span class='warning'>There is no coin in this machine.</span>")
+				return
+			if(istype(user, /mob/living/silicon))
+				to_chat(user, "<span class='warning'>You lack hands.</span>")
+				return
+			to_chat(user, "<span class='notice'>You remove [coin] from [src].</span>")
+			user.put_in_hands(coin)
+			coin = null
+			. = TRUE
+		if("change")
+			. = TRUE
+			give_change(user)
+		if("vend")
+			if(!vend_ready)
```

Comment:
Can you make this into a proc? That's a lot of lines

Reply:
Yeah I can try, I've been trying to avoid refactoring everything I'm touching just because its a massive amount of code

Reply:
should be done in next commit